File: java/org.financer.client.javafx/src/test/java/org/financer/client/javafx/main/VariableTransactionTest.java
package org.financer.client.javafx.main;

import com.jfoenix.controls.JFXButton;
import javafx.scene.control.Button;
import javafx.scene.control.TableView;
import javafx.scene.control.TextField;
import javafx.scene.input.KeyCode;
import org.financer.client.domain.model.category.Category;
import org.financer.client.domain.model.category.CategoryRoot;
import org.financer.client.domain.model.transaction.Transaction;
import org.financer.client.domain.model.transaction.VariableTransaction;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.shared.domain.model.value.objects.Amount;
import org.financer.util.collections.TreeUtil;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.Comparator;
import java.util.Set;

@SuppressWarnings("WeakerAccess")
@Tag("integration")
public class VariableTransactionTest extends AbstractFinancerApplicationTest {

    @BeforeEach
    public void setUpEach() throws Exception {
        super.setUpEach();
    }

    @Test
    public void testAddTransaction() {
        final VariableTransaction transaction = variableTransaction();
        register(user(), password());
        addCategory(variableCategory());
        addVariableTransaction(transaction);
        clickOn((Button) find("#refreshTransactionsBtn"));
        sleep(1000);

        Assertions.assertNotNull(clickOn(formatter.format(transaction.getAmount())));

        CategoryRoot categoryRoot = LocalStorageImpl.getInstance().readObject("categories");
        Set<Transaction> transactions = ((Category) TreeUtil.getByValue(categoryRoot, transaction.getCategory(),
                Comparator.comparingLong(Category::getId))).getTransactions();
        Assertions.assertEquals(1, transactions.size());
        for (Transaction _transaction : transactions) {
            Assertions.assertTrue(_transaction instanceof VariableTransaction);
            Assertions.assertTrue(_transaction.getId() > 0);
        }
    }

    @Test
    public void testEditTransaction() {
        final VariableTransaction transaction = variableTransaction();
        register(user(), password());
        addCategory(variableCategory());
        addVariableTransaction(transaction);
        clickOn(formatter.format(transaction.getValueDate()));
        clickOn((Button) find("#editTransactionBtn"));
        final Amount amount = transaction.getAmount().divide(2);
        TextField amountTextField = find("#amountTextField");
        clickOn(amountTextField);
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE);
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE);
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE);
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE);
        write(Double.toString(amount.getAmount()));
        confirmDialog();
        clickOn((Button) find("#refreshTransactionsBtn"));
        sleep(MEDIUM_SLEEP);

        Assertions.assertNotNull(clickOn(formatter.format(amount)));
        CategoryRoot categoryRoot = LocalStorageImpl.getInstance().readObject("categories");
        Set<Transaction> transactions = ((Category) TreeUtil.getByValue(categoryRoot, transaction.getCategory(),
                Comparator.comparingLong(Category::getId))).getTransactions();
        Assertions.assertEquals(1, transactions.size());
        for (Transaction _transaction : transactions) {
            Assertions.assertTrue(_transaction instanceof VariableTransaction);
            Assertions.assertTrue(_transaction.getId() > 0);
            Assertions.assertEquals(amount, _transaction.getAmount());
        }
    }

    @Test
    public void testDeleteTransaction() {
        final VariableTransaction transaction = variableTransaction();
        register(user(), password());
        addCategory(variableCategory());
        addVariableTransaction(transaction);
        clickOn((Button) find("#refreshTransactionsBtn"));
        sleep(MEDIUM_SLEEP);
        clickOn(formatter.format(transaction.getValueDate()));
        clickOn((JFXButton) find("#deleteTransactionBtn"));
        confirmDialog();
        sleep(MEDIUM_SLEEP);
        Assertions.assertEquals(0, ((TableView) find("#transactionsTableView")).getItems().size());
        CategoryRoot categoryRoot = LocalStorageImpl.getInstance().readObject("categories");
        Assertions.assertEquals(0, ((Category) TreeUtil.getByValue(categoryRoot, transaction.getCategory(),
                Comparator.comparingLong(Category::getId))).getTransactions().size());
    }
}

File: java/org.financer.client.javafx/src/test/kotlin/org/financer/client/javafx/main/VariableTransactionTest.kt
package org.financer.client.javafx.main

import com.jfoenix.controls.JFXButton
import javafx.scene.Node
import javafx.scene.control.Button
import javafx.scene.control.TableView
import javafx.scene.control.TextField
import javafx.scene.input.KeyCode
import org.financer.client.domain.model.category.Category
import org.financer.client.domain.model.category.CategoryRoot
import org.financer.client.domain.model.transaction.VariableTransaction
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.util.collections.TreeUtil
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import java.util.*

@Tag("integration")
class VariableTransactionTest : AbstractFinancerApplicationTest() {
    @BeforeEach
    @Throws(Exception::class)
    public override fun setUpEach() {
        super.setUpEach()
    }

    @Test
    fun testAddTransaction() {
        val transaction = variableTransaction()
        register(user(), password())
        addCategory(variableCategory())
        addVariableTransaction(transaction!!)
        clickOn(find<Node>("#refreshTransactionsBtn") as Button)
        sleep(1000)
        Assertions.assertNotNull(clickOn(formatter!!.format(transaction.amount)))
        val categoryRoot = LocalStorageImpl.getInstance().readObject<CategoryRoot>("categories")
        val transactions = (TreeUtil.getByValue(categoryRoot, transaction.category,
                Comparator.comparingLong(Category::id)) as Category).transactions
        Assertions.assertEquals(1, transactions.size)
        for (_transaction in transactions) {
            Assertions.assertTrue(_transaction is VariableTransaction)
            Assertions.assertTrue(_transaction.id > 0)
        }
    }

    @Test
    fun testEditTransaction() {
        val transaction = variableTransaction()
        register(user(), password())
        addCategory(variableCategory())
        addVariableTransaction(transaction!!)
        clickOn(formatter!!.format(transaction.valueDate!!))
        clickOn(find<Node>("#editTransactionBtn") as Button)
        val amount = transaction.amount.divide(2.0)
        val amountTextField = find<TextField>("#amountTextField")
        clickOn(amountTextField)
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE)
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE)
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE)
        press(KeyCode.BACK_SPACE).release(KeyCode.BACK_SPACE)
        write(java.lang.Double.toString(amount.amount))
        confirmDialog()
        clickOn(find<Node>("#refreshTransactionsBtn") as Button)
        sleep(AbstractFinancerApplicationTest.Companion.MEDIUM_SLEEP.toLong())
        Assertions.assertNotNull(clickOn(formatter!!.format(amount)))
        val categoryRoot = LocalStorageImpl.getInstance().readObject<CategoryRoot>("categories")
        val transactions = (TreeUtil.getByValue(categoryRoot, transaction.category,
                Comparator.comparingLong(Category::id)) as Category).transactions
        Assertions.assertEquals(1, transactions.size)
        for (_transaction in transactions) {
            Assertions.assertTrue(_transaction is VariableTransaction)
            Assertions.assertTrue(_transaction.id > 0)
            Assertions.assertEquals(amount, _transaction.amount)
        }
    }

    @Test
    fun testDeleteTransaction() {
        val transaction = variableTransaction()
        register(user(), password())
        addCategory(variableCategory())
        addVariableTransaction(transaction!!)
        clickOn(find<Node>("#refreshTransactionsBtn") as Button)
        sleep(AbstractFinancerApplicationTest.Companion.MEDIUM_SLEEP.toLong())
        clickOn(formatter!!.format(transaction.valueDate!!))
        clickOn(find<Node>("#deleteTransactionBtn") as JFXButton)
        confirmDialog()
        sleep(AbstractFinancerApplicationTest.Companion.MEDIUM_SLEEP.toLong())
        Assertions.assertEquals(0, (find<Node>("#transactionsTableView") as TableView<*>).items.size)
        val categoryRoot = LocalStorageImpl.getInstance().readObject<CategoryRoot>("categories")
        Assertions.assertEquals(0, (TreeUtil.getByValue(categoryRoot, transaction.category,
                Comparator.comparingLong(Category::id)) as Category).transactions.size)
    }
}